ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/result/impl/RangeActionActivationResultImpl.java

```diff
@@ -131,48 +125,51 @@ public Set<RangeAction<?>> getActivatedRangeActions(State state) {
     }
 
     @Override
-    public synchronized double getOptimizedSetpoint(RangeAction<?> rangeAction, State state) {
-        if (!(memoizedSetpointPerStatePerRangeAction.containsKey(rangeAction) && memoizedSetpointPerStatePerRangeAction.get(rangeAction).containsKey(state))) {
-            // find all range actions on same network elements
-            Set<RangeAction<?>> correspondingRa = elementaryResultMap.keySet().stream()
-                .filter(ra -> ra.getId().equals(rangeAction.getId()) || (ra.getNetworkElements().equals(rangeAction.getNetworkElements())))
-                .collect(Collectors.toSet());
-
-            if (correspondingRa.isEmpty()) {
-                throw new FaraoException(format("range action %s is not present in the result", rangeAction.getName()));
-            }
-
-            Optional<RangeAction<?>> lastActivatedRaOpt = correspondingRa.stream()
-                .max(Comparator.comparingInt(ra -> {
-                    Optional<State> lastActivation = elementaryResultMap.get(ra).getSetpointAndLastActivation(state).getRight();
-                    return lastActivation.isPresent() ? lastActivation.get().getInstant().getOrder() : -1;
-                }));
-
-            if (lastActivatedRaOpt.isPresent()) {
-                memoizedSetpointPerStatePerRangeAction.computeIfAbsent(rangeAction, ra -> new HashMap<>());
-                memoizedSetpointPerStatePerRangeAction.get(rangeAction).put(state, elementaryResultMap.get(lastActivatedRaOpt.get()).getSetpoint(state));
+    public double getOptimizedSetpoint(RangeAction<?> rangeAction, State state) {
+        computeSetpointsPerStatePerPst();
+        String networkElementsIds = concatenateNetworkElementsIds(rangeAction);
+        // if at least one elementary result is on the correct network elements, find the right state to get the setpoint
+        // else return the reference setpoint
+        if (setpointPerStatePerPstId.containsKey(networkElementsIds)) {
+            Map<State, Double> setPointPerState = setpointPerStatePerPstId.get(networkElementsIds);
+            // if an elementary result is defined for the network element and state, return it
+            // else find a previous state with an elementary result
+            // if none are found, return reference setpoint
+            if (setPointPerState.containsKey(state)) {
```

Comment:
can you create a recursive method instead? something like:
```suggestion
            Double lastSetpoint = getLastSetpoint(setPointPerState, state);
            if (lastSetPoint != null) {
                return lastSetPoint;
            }
            
            ...
            
            Double getLastSetpoint(Map setPointPerState, State state) {
                if (state == null) {
                    return null;
                }
                if (setPointPerState.containsKey(state)) {
                    return setPointPerState.get(state);
                }
                return getLastSetpoint(setPointPerState, getPreviousState(state).orElse(null));
            }
         
```